### PR TITLE
Enable systemd-cryptenroll to support pcr literals on the command line.

### PR DESCRIPTION
**Pull request by flixman**
THIS IS WORK IN PROGRESS. 
Current implementation of tpm2-related tools in systemd-cryptenroll support specifying a PCR index, but not giving a literal hash value. This PR provides such functionality. The idea is to reuse as much code as possible, so the specification of the registers would go like
```
 --tpm2-pcrs=7+11:sha256=<hash>+14
```

### CODE REVIEW

File: src/shared/tpm2-util.c [hunk 1]
```diff
@@ -3797,45 +3819,85 @@ char *tpm2_pcr_mask_to_string(uint32_t mask) {
         return TAKE_PTR(s);
 }
 
-int tpm2_pcr_mask_from_string(const char *arg, uint32_t *ret_mask) {
-        uint32_t mask = 0;
+int hex_to_bytes(uint8_t bytes[SHA256_DIGEST_SIZE], const char *hex) {
```
Comment (poettering) [inline src/shared/tpm2-util.c]: unhexmem()

File: src/creds/creds.c [hunk 2]
```diff
@@ -45,6 +45,8 @@ static int arg_newline = -1;
 static sd_id128_t arg_with_key = _CRED_AUTO;
 static const char *arg_tpm2_device = NULL;
 static uint32_t arg_tpm2_pcr_mask = UINT32_MAX;
+static uint32_t arg_tpm2_literal_mask = UINT32_MAX;
+static uint8_t arg_tpm2_pcr_literal[24][SHA256_DIGEST_SIZE];
```
Comment (poettering) [inline src/creds/creds.c]: as mentioned elsewhere. Please just make this a `Hashmap` object, keyed by a combination of pcr number and algorithm

File: src/shared/tpm2-util.c [hunk 3]
```diff
@@ -3177,15 +3181,30 @@ int tpm2_seal(const char *device,
...
+          size_t count = 0;
+          for (int i=0; i<24; i++) {
+                if ((literal_pcr_mask >> i) & 1) {
+                        memcpy_safe(hash_pcr_values[count].buffer, hash_pcr_literal[i], SHA256_DIGEST_SIZE);
+                }
+                /* update the count if this bit is set in the aggregated mask */
+                count += ((agg_pcr_mask >> i) & 1);
+          }
+
+        }
```
Comment (poettering) [inline src/shared/tpm2-util.c:3221]: indentation off. please follow coding style

File: src/shared/tpm2-util.c [hunk 4]
```diff
@@ -3177,15 +3181,30 @@ int tpm2_seal(const char *device,
...
                         return r;
         }
 
+        /* In case there are literal hashes passed, overwrite the existing values */
+        if (literal_pcr_mask) {
+          size_t count = 0;
+          for (int i=0; i<24; i++) {
+                if ((literal_pcr_mask >> i) & 1) {
+                        memcpy_safe(hash_pcr_values[count].buffer, hash_pcr_literal[i], SHA256_DIGEST_SIZE);
+                }
```
Comment (poettering) [inline src/shared/tpm2-util.c:3219]: no {} around single-line if blocks, see coding style

File: src/shared/tpm2-util.c [hunk 5]
```diff
@@ -3797,45 +3819,85 @@ char *tpm2_pcr_mask_to_string(uint32_t mask) {
...
-                n = r;
-                SET_BIT(mask, n);
+                /* detect if this is (potentially) a pcr index or a pcr literal*/
+                r2 = strchr(pcr, ':');
+
+                /* if this is a pcr literal, try to parse it */
+                if (r2) {
+                        unsigned pcr_idx;
+                        char pcr_hex[2*SHA256_DIGEST_SIZE+1];
+                        r = sscanf(pcr, "%u:sha256=%s", &pcr_idx, pcr_hex);
```
Comment (poettering) [inline src/shared/tpm2-util.c]: this should be reasonably agnostic to hash functions. i.e. do not hardcode sha256 here. 

also, please don't parse with sscanf(). Use extract_first_word(), or or startswith() or so.

File: src/shared/tpm2-util.c [hunk 6]
```diff
@@ -3797,45 +3819,85 @@ char *tpm2_pcr_mask_to_string(uint32_t mask) {
...
+                        if (r != 2)
+                                return log_error_errno(-EINVAL, "Failed to parse specified PCR literal: %s", pcr);
```
Comment (poettering) [inline src/shared/tpm2-util.c]: use `return log_error_errno(SYNTHETIC_ERRNO(EINVAL), …`

File: src/shared/tpm2-util.c [hunk 7]
```diff
@@ -3797,45 +3819,85 @@ char *tpm2_pcr_mask_to_string(uint32_t mask) {
...
+
+                        r = hex_to_bytes(ret_literal[pcr_idx], pcr_hex);
+                        if (r < 0) {
+                                return log_error_errno(r, "Failed to parse specified PCR literal: %s", pcr);
+                        }
```
Comment (poettering) [inline src/shared/tpm2-util.c:3968]: no {} around single line if blocks

File: src/shared/tpm2-util.c [hunk 8]
```diff
@@ -3797,45 +3819,85 @@ char *tpm2_pcr_mask_to_string(uint32_t mask) {
...
+
+                        SET_BIT(literal_mask, pcr_idx);
+                } else
+                /* if this is a pcr index, parse it */
+                {
```
Comment (poettering) [inline src/shared/tpm2-util.c]: `} else {` please

File: src/shared/tpm2-util.c [hunk 9]
```diff
@@ -4200,11 +4264,22 @@ Tpm2Support tpm2_support(void) {
...
 }
 
-int tpm2_parse_pcr_argument(const char *arg, uint32_t *mask) {
-        uint32_t m;
+void copy_literal_sha256_pcr(uint8_t dest[][SHA256_DIGEST_SIZE], uint8_t src[][SHA256_DIGEST_SIZE], uint32_t bitmask) {
+            for (int i = 0; i < 24; i++) {
+                if ((bitmask >> i) & 1) {
+                        memcpy_safe(dest[i], src[i], SHA256_DIGEST_SIZE);
+                }
+        }
```
Comment (poettering) [inline src/shared/tpm2-util.c]: indentation fucke,and no {} around signle line if blocks

File: src/test/test-tpm2.c [hunk 10]
```diff
@@ -4,37 +4,74 @@
...
+        test_tpm2_pcr_from_string_one("7+application-support", 8388736, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("8+boot-loader-code", 272, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("6+boot-loader-code,44", 0, literal_mask, pcr_literal, -EINVAL);
+        test_tpm2_pcr_from_string_one("7,shim-policy,4", 16528, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("sysexts,shim-policy+kernel-boot", 26624, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("sysexts,shim+kernel-boot", 0, literal_mask, pcr_literal, -EINVAL);
+        test_tpm2_pcr_from_string_one("sysexts+17+23", 8527872, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("debug+24", 0, literal_mask, pcr_literal, -EINVAL);
+
+        hex = (char *)"3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969";
```
Comment (poettering) [inline src/test/test-tpm2.c]: why cast the const away? just make `hex` const?

File: src/test/test-tpm2.c [hunk 11]
```diff
@@ -4,37 +4,74 @@
...
+        pcr_idx = 11;
+        sprintf(pcr_string, "0,%u:sha256=%s", pcr_idx, hex);
+        memset(pcr_literal, 0, 24*SHA256_DIGEST_SIZE);
```
Comment (poettering) [inline src/test/test-tpm2.c]: zero() is your friend

File: src/test/test-tpm2.c [hunk 12]
```diff
@@ -4,37 +4,74 @@
...
+        test_tpm2_pcr_from_string_one("6+boot-loader-code,44", 0, literal_mask, pcr_literal, -EINVAL);
+        test_tpm2_pcr_from_string_one("7,shim-policy,4", 16528, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("sysexts,shim-policy+kernel-boot", 26624, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("sysexts,shim+kernel-boot", 0, literal_mask, pcr_literal, -EINVAL);
+        test_tpm2_pcr_from_string_one("sysexts+17+23", 8527872, literal_mask, pcr_literal, 0);
+        test_tpm2_pcr_from_string_one("debug+24", 0, literal_mask, pcr_literal, -EINVAL);
+
+        hex = (char *)"3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969";
+        pcr_idx = 11;
+        sprintf(pcr_string, "0,%u:sha256=%s", pcr_idx, hex);
```
Comment (poettering) [inline src/test/test-tpm2.c]: no sprintf() in our codebase please. xsprintf() at least. or better: asprintf()

File: src/shared/tpm2-util.c [hunk 13]
```diff
@@ -3797,12 +3819,13 @@ char *tpm2_pcr_mask_to_string(uint32_t mask) {
         return TAKE_PTR(s);
 }
 
-int tpm2_pcr_mask_from_string(const char *arg, uint32_t *ret_mask) {
+int tpm2_pcr_from_string(const char *arg, uint32_t *ret_mask, uint32_t *ret_literal_mask, uint8_t *ret_literal[]) {
```
Comment (ddstreet) [inline src/shared/tpm2-util.c]: let's not have to manage two sets of pcr selections, *plus* the signed pcr selections?

**Review by ddstreet — COMMENTED** [review]
I'm also working on adding this, BTW....

**Comment by flixman** [comment]
@poettering @ddstreet I have been traveling for the last two days. After your remark on the issue I have moved everything to hashmap and I am applying the coding style. My apologies for the (lack of) quality of this draft, but its purpose was only to show the proposal to get any architectural concerns. I hope I will finish this in the coming few days (I am still on the go, with limited time :-/).

@ddstreet: on your remark about tpm2_pcr_from_string: I am currently changing the signature of tpm2_pcr_from_string so that it does not return the literals mask and the values themselves, but only a <bank, pcrs> hashmap, where _pcrs_ is another <index, digest>, in which _digest_ is a TPM2B_DIGEST. Additionally, I am providing a function that, given the pcrs hashmap, will return the mask for those literals.

@poettering: I will incorporate your remarks in the coming commits. Thank you very much! I am having a bit of a problem accessing a hashmap whose values are also hashmaps, but I will update the PR as soon as I get around it.

**Comment by ddstreet** [comment]
I don't mean to step on your work, but I have #28398 which I think is a more comprehensive approach.

**Comment by flixman** [comment]
@ddstreet Good! then I stop on that. Good luck!

**Pull request closed without merging**